magicblock-processor/src/scheduler/locks.rs

```diff
@@ -0,0 +1,109 @@
+//! Fast, in-memory account locking primitives for the multi-threaded scheduler.
+//!
+//! This version uses a single `u64` bitmask to represent the entire lock state,
+//! including read locks, write locks, and contention, for maximum efficiency.
+
+use std::{cell::RefCell, collections::VecDeque, rc::Rc};
+
+use rustc_hash::FxHashMap;
+use solana_pubkey::Pubkey;
+
+// A bitmask representing the lock state.
+// - MSB: Write lock flag.
+// - Remaining bits: Read locks for each executor.
+type ReadWriteLock = u64;
+
+/// Unique identifier for a transaction executor worker.
+pub(crate) type ExecutorId = u32;
+
+/// Unique identifier for a transaction to be scheduled.
+///
+/// NOTE: the type is specifically set to u64, so that it
+/// will be statistically impossible to overlow it within
+/// next few millenia, given the indended use case of the
+/// type as a tagging counter for incoming transactions.
+pub(super) type TransactionId = u64;
+
+/// A shared, mutable reference to an `AccountLock`.
+pub(super) type RcLock = Rc<RefCell<AccountLock>>;
+
+/// In-memory cache of account locks.
+pub(super) type LocksCache = FxHashMap<Pubkey, RcLock>;
+/// A map from a blocked transaction to the executor that holds the conflicting lock.
+pub(super) type TransactionContention = FxHashMap<TransactionId, ExecutorId>;
+/// A map from account's pubkey to all transaction, that are contending to acquire its lock.
+pub(super) type AccountContention = FxHashMap<Pubkey, VecDeque<TransactionId>>;
+
+/// The maximum number of concurrent executors supported by the bitmask.
+/// One bit is reserved for the write flag.
+pub(super) const MAX_SVM_EXECUTORS: u32 = ReadWriteLock::BITS - 1u32;
+
+/// The bit used to indicate a write lock is held. This is the most significant bit.
+const WRITE_BIT_MASK: u64 = 1u64 << (ReadWriteLock::BITS - 1u32);
+/// The starting transaction ID, follows the max possible value for ExecutorId
+pub(crate) const INIT_TRANSACTION_ID: TransactionId =
+    MAX_SVM_EXECUTORS as TransactionId;
+
+/// A read/write lock on a single Solana account, represented by a `u64` bitmask.
+#[derive(Default, Debug)]
+pub(super) struct AccountLock {
+    rw: ReadWriteLock,
+}
+
+impl AccountLock {
+    /// Attempts to acquire a write lock. Fails if any other lock is held.
+    #[inline]
+    pub(super) fn write(
+        &mut self,
+        executor: ExecutorId,
+    ) -> Result<(), ExecutorId> {
+        if self.rw != 0 {
+            // If the lock is held, `trailing_zeros()` will return the index of the
+            // least significant bit that is set. This corresponds to the ID of the
```

Comment:
I think this part

>  This corresponds to the ID of the executor that holds the lock.

Should be moved in docs of the function. Something like

`/// Attempts to acquire a write lock. Fails returning id of blocking Executor`

Reply:
Done 908e788c5dff6a9affe56d23efa0b3a6bd0d4f43